biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Indexer.java

```diff
@@ -130,12 +132,29 @@ private Repository readRepository() throws Exception {
 		P2Impl p2 = new P2Impl(client, this.url, Processor.getExecutor());
 		List<Artifact> artifacts = p2.getArtifacts();
 		List<Promise<Resource>> fetched = new ArrayList<>(artifacts.size());
-		Set<URI> visited = new HashSet<>(artifacts.size());
+		Set<BundleId> visited = new HashSet<>(artifacts.size());
+
+		final Map<BundleId,Resource> existing = new HashMap<>();
+		if (getBridge() != null) {
```

Comment:
It's a refresh in that it removes artefacts which are nolonger in the repo and adds artefacts which have been added to the repo since it was last indexed. It's only a no-op for artefacts which it validates as unchanged.

As for whether that validation is currently sufficient ... I happily accept that it's not, but I think "horribly broken" is a bit exaggerated. With hashes considered as discussed in the issue comments I think this approach is fine.

If a refresh has to build the index from scratch then there's no point in even keeping the index file in the cache for Eclipse users. There is no point in a cache that is discarded as soon as you start the IDE.

There's also no point in "refreshing" a resource if we can validate that it's unchanged. In this case the refresh *should be* a no-op. If we suspect the index has become corrupted or that the previous indexing failed for some reason then yes we need a hard reset. But I do not accept that this needs to be done every time I open the IDE just in case. Just delete the cache if the cache is broken.